event-stream-processing/src/parsers/document-index.parser.ts

```diff
@@ -34,13 +35,24 @@ export class DocumentIndexParser implements Parser {
       throw new Error('@timestamp field value has not been defined');
     }
     const tsMomement = moment(timestamp);
-
-    const indexFormat: string = index.replace(/\<\%\=[^\=]+=\%\>/gm, (match: string) => {
-      if (match.startsWith('<%=')) {
-        return tsMomement.format(match.substring(3, match.length - 3));
-      }
-      throw new Error(`Unexpected formatting: ${match}`);
-    });
-    document.index = indexFormat;
+    if (!app) {
+      const indexFormat: string = index.replace(/\<\%\=[^\=]+=\%\>/gm, (match: string) => {
+        if (match.startsWith('<%=')) {
+          return tsMomement.format(match.substring(3, match.length - 3));
+        }
+        throw new Error(`Unexpected formatting: ${match}`);
+      });
+      document.index = indexFormat;
+    }
+    if (app) {
+      const indexFormatWithApp = index.replace(/\<!\=[^\=]+=!\>/gm, app);
+      const indexFormat: string = indexFormatWithApp.replace(/\<\%\=[^\=]+=\%\>/gm, (match: string) => {
+        if (match.startsWith('<%=')) {
+          return tsMomement.format(match.substring(3, match.length - 3));
+        }
+        throw new Error(`Unexpected formatting: ${match}`);
+      });
+      document.index = indexFormat;
+    }
```

Comment:
Problems:
- Duplicates code
- Should be able to generically substitute any data field
- It does not fail if substitution for labels.application requested but the field does not exist
- No unit tests

Reply:
@andrwils